src/github.com/getlantern/enproxy/conn_test.go

```diff
@@ -62,18 +60,80 @@ func TestBadBuffered(t *testing.T) {
 	doTestBad(true, t)
 }
 
+func TestIdle(t *testing.T) {
+	idleTimeout := 100 * time.Millisecond
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	_, err = Dial(httpAddr, &Config{
+		DialProxy: func(addr string) (net.Conn, error) {
+			return net.Dial("tcp", proxyAddr)
+		},
+		NewRequest:  newRequest,
+		IdleTimeout: idleTimeout,
+	})
+	if assert.NoError(t, err, "Dialing should have succeeded") {
+		time.Sleep(idleTimeout * 2)
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+	}
+}
+
+// This test stimulates a connection leak as seen in
+// https://github.com/getlantern/lantern/issues/2174.
+func TestHTTPRedirect(t *testing.T) {
+	startProxy(t)
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			Dial: func(network, addr string) (net.Conn, error) {
+				return Dial(addr, &Config{
+					DialProxy: func(addr string) (net.Conn, error) {
+						return net.Dial("tcp", proxyAddr)
+					},
+					NewRequest: newRequest,
+				})
+			},
+			DisableKeepAlives: true,
+		},
+	}
+
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
+	resp, err := client.Head("http://www.facebook.com")
+	if assert.NoError(t, err, "Head request to facebook should have succeeded") {
+		resp.Body.Close()
+	}
+
+	assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
+}
+
 func doTestPlainText(buffered bool, t *testing.T) {
 	startServers(t)
 
+	_, counter, err := fdcount.Matching("TCP")
+	if err != nil {
+		t.Fatalf("Unable to get fdcount: %v", err)
+	}
+
 	conn, err := prepareConn(httpAddr, buffered, false, t)
 	if err != nil {
 		t.Fatalf("Unable to prepareConn: %s", err)
 	}
-	defer conn.Close()
+	defer func() {
+		err := conn.Close()
+		assert.Nil(t, err, "Closing conn should succeed")
+		assert.NoError(t, counter.AssertDelta(2), "All file descriptors except the connection from proxy to destination site should have been closed")
```

Comment:
Hello @aranhoide, does this error look familiar to you?

I've got the same error on Mac (go1.4.1) so it could not be a Linux-only problem.

```
2015/01/30 10:51:27 http: TLS handshake error from 127.0.0.1:53729: EOF
enproxy: 
---- Enproxy Connections----
Open:                         0
Closing:                      0
Blocked on Closing:           0
Blocked on Read:              0
Reading:                      0
Reading Finishing:            0
Blocked on Write:             0
Writing:                      0
  Selecting:                  0
  Writing:                    0
    Write Pipe Open:          0
    Request Pending:          0
      Submitting Req.:        0
      Processing Req.:        0
        Posting Req. Fin:     0
        Posting Resp:         0       
        Dialing First:        0
        Redialing:            0
    Doing Write:              0
  Posting Response:           0
  Writing Empty:              0
  Finishing Body:             0
  Finishing:                  0
Requesting:                   0
Requesting Finishing:         0

--- FAIL: TestPlainTextStreaming (0.75s)
        Location:       conn_test.go:131
        Error:          No error is expected but got Unexpected TCP file descriptor count. Expected 2, have 1.

                        Removed file descriptors
                        -----------------------------
                        enproxy.t 1155 reventlov   10u    IPv4 0x6510f8cd1abdeec5       0t0     TCP localhost:53730->localhost:53731 (CLOSE_WAIT)

                        New file descriptors
                        -----------------------------
                        enproxy.t 1155 reventlov   12u    IPv4 0x6510f8cd1f1a8b85       0t0     TCP localhost:53733->localhost:53719 (ESTABLISHED)
                        enproxy.t 1155 reventlov   13u    IPv4 0x6510f8cd1f1bdd25       0t0     TCP localhost:53719->localhost:53733 (ESTABLISHED)
        Messages:       All file descriptors except the connection from proxy to destination site should have been closed
```

I'll continue studying the code, it would be great training fixing this.